src/BSON/Iterator.c

```diff
@@ -263,10 +298,92 @@ static HashTable* php_phongo_iterator_get_properties(phongo_compat_object_handle
 	return php_phongo_iterator_get_properties_hash(object, false);
 }
 
+/* Iterator handlers */
+static void php_phongo_iterator_it_dtor(zend_object_iterator* iter)
+{
+	zval_ptr_dtor(&iter->data);
+}
+
+static int php_phongo_iterator_it_valid(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	return intern->valid ? SUCCESS : FAILURE;
+}
+
+static zval* php_phongo_iterator_it_get_current_data(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	return php_phongo_iterator_get_current(intern);
```

Comment:
Unlike the `current()` method itself, there's no need to copy/deref the zval we're returning here. Another notable difference to `current()` is that we're leaving an `IS_UNDEF` zval as-is, whereas `current` returns `null` in those cases. I don't think it makes a significant difference, and I'm not sure why `Cursor::current()` (which inspired the BSON iterator) returns `null` for an undef zval. I know there are some weird shenanigans going on when returning a `IS_UNDEF`, but I don't think this will ever matter in practice, as having that value returned indicates that something went seriously wrong during the iteration and there should've been an exception anyways.